### PR TITLE
[core] implement adjustProjection in PartialUpdateMergeFunction

### PR DESCRIPTION
**Pull request by LsomeYeah**
### Purpose

When user provides a projection, we should expand the projection with the necessary sequence fields. 
And the FieldSequenceGroup should be adjusted if the projection is applied.

Linked issue: close #xxx

### Tests

UT case : add a test `testAdjustProjection` in `PartialUpdateMergeFunctionTest`.
IT case : enrich the content of `testSequenceGroup` in `PartialUpdateITCase`

### CODE REVIEW

File: paimon-core/src/main/java/org/apache/paimon/mergetree/compact/PartialUpdateMergeFunction.java [hunk 1]
```diff
@@ -203,7 +231,35 @@ public AdjustedProjection adjustProjection(@Nullable int[][] projection) {
             if (fieldSequences.isEmpty()) {
                 return new AdjustedProjection(projection, null);
             }
-            // TODO implement this, just keep required fields and adjust fieldSequences too.
+
+            if (projection != null) {
+                int[] projIndexs = Projection.of(projection).toTopLevelIndexes();
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/mergetree/compact/PartialUpdateMergeFunction.java:236]:
```
            if (projection == null) {
                return new AdjustedProjection(null, null);
            }
            LinkedHashSet<Integer> extraFields = new LinkedHashSet<>();
            int[] topProjects = Projection.of(projection).toTopLevelIndexes();
            for (int index : topProjects) {
                SequenceGenerator generator = fieldSequences.get(index);
                if (generator != null) {
                    extraFields.add(generator.index());
                }
            }

            int[] allProjects = Stream.concat(
                    Arrays.stream(topProjects).boxed(),
                    extraFields.stream()).mapToInt(Integer::intValue).toArray();

            int[][] pushdown = Projection.of(allProjects).toNestedIndexes();
            int[][] outer = Projection.of(IntStream.range(0, topProjects.length).toArray()).toNestedIndexes();
            return new AdjustedProjection(pushdown, outer);
```

File: paimon-core/src/main/java/org/apache/paimon/mergetree/compact/PartialUpdateMergeFunction.java [hunk 2]
```diff
@@ -190,11 +193,36 @@ private Factory(Options options, RowType rowType) {
         @Override
         public MergeFunction<KeyValue> create(@Nullable int[][] projection) {
             List<DataType> fieldTypes = tableTypes;
+            Map<Integer, SequenceGenerator> projFieldSequences = new HashMap<>();
+
             if (projection != null) {
                 fieldTypes = Projection.of(projection).project(tableTypes);
+                int[] fieldOldIndex = Projection.of(projection).toTopLevelIndexes();
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/mergetree/compact/PartialUpdateMergeFunction.java:200]: naming: fieldOldIndex  -> projects

File: paimon-core/src/main/java/org/apache/paimon/mergetree/compact/PartialUpdateMergeFunction.java [hunk 3]
```diff
@@ -190,11 +193,36 @@ private Factory(Options options, RowType rowType) {
         @Override
         public MergeFunction<KeyValue> create(@Nullable int[][] projection) {
             List<DataType> fieldTypes = tableTypes;
+            Map<Integer, SequenceGenerator> projFieldSequences = new HashMap<>();
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/mergetree/compact/PartialUpdateMergeFunction.java:196]: naming: projectedFieldSequences

File: paimon-core/src/main/java/org/apache/paimon/mergetree/compact/PartialUpdateMergeFunction.java [hunk 4]
```diff
@@ -190,11 +193,36 @@ private Factory(Options options, RowType rowType) {
...
+
             if (projection != null) {
                 fieldTypes = Projection.of(projection).project(tableTypes);
+                int[] fieldOldIndex = Projection.of(projection).toTopLevelIndexes();
+                Map<Integer, Integer> indexMap = new HashMap<>();
+                for (int i = 0; i < fieldOldIndex.length; i++) {
+                    indexMap.put(fieldOldIndex[i], i);
+                }
+                for (int index = 0; index < fieldOldIndex.length; index++) {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/mergetree/compact/PartialUpdateMergeFunction.java:205]: loop `fieldSequences` like this:
```
fieldSequences.forEach((field, sequence) -> {
    int newField = indexMap.get(field);
    checkArguement(newField != -1);
    int newSequenceId = indexMap.get(sequence.index());
    projectedFieldSequences.put(newField, new SequenceGenerator(newSequenceId, ...));
}); 
```

File: paimon-core/src/main/java/org/apache/paimon/mergetree/compact/PartialUpdateMergeFunction.java [hunk 5]
```diff
@@ -190,11 +193,36 @@ private Factory(Options options, RowType rowType) {
...
+                    int oldIndex = fieldOldIndex[index];
+                    if (fieldSequences.get(oldIndex) != null) {
+                        int oldKeyIndex = fieldSequences.get(oldIndex).index();
+                        int keyIndex = indexMap.get(oldKeyIndex);
+                        if (!projFieldSequences.containsKey(keyIndex)) {
+                            SequenceGenerator newSequenceGen =
+                                    new SequenceGenerator(keyIndex, fieldTypes.get(keyIndex));
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/mergetree/compact/PartialUpdateMergeFunction.java:212]: keyIndex is not new index of oldIndex.
There should be a bug! You should add tests to cover.

File: paimon-core/src/test/java/org/apache/paimon/mergetree/compact/PartialUpdateMergeFunctionTest.java [hunk 6]
```diff
@@ -95,6 +95,38 @@ public void testSequenceGroupRepeatDefine() {
                 .hasMessageContaining("is defined repeatedly by multiple groups");
     }
 
+    @Test
+    public void testAdjustProjection() {
```
Comment (JingsongLi) [inline paimon-core/src/test/java/org/apache/paimon/mergetree/compact/PartialUpdateMergeFunctionTest.java:99]: This test is too simple, you should cover more.

File: paimon-core/src/main/java/org/apache/paimon/mergetree/compact/PartialUpdateMergeFunction.java (hunk 1 above)
Comment (LsomeYeah) [reply paimon-core/src/main/java/org/apache/paimon/mergetree/compact/PartialUpdateMergeFunction.java:236]: There may be some sequence fields in `topProjects`, we don't need to add these fields to `extraFields`.

File: paimon-core/src/main/java/org/apache/paimon/mergetree/compact/PartialUpdateMergeFunction.java [hunk 7]
```diff
@@ -190,11 +195,36 @@ private Factory(Options options, RowType rowType) {
...
+                Map<Integer, Integer> indexMap = new HashMap<>();
+                for (int i = 0; i < projects.length; i++) {
+                    indexMap.put(projects[i], i);
+                }
+                fieldSequences.forEach(
+                        (field, sequence) -> {
+                            int newField = indexMap.getOrDefault(field, -1);
+                            if (newField != -1) {
+                                int newSequenceId = indexMap.getOrDefault(sequence.index(), -1);
+                                if (newSequenceId != -1) {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/mergetree/compact/PartialUpdateMergeFunction.java:212]: newSequenceId == -1 should throws a exception, you should add test for this.

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**